test/e2e/upgrades/ingress.go

```diff
@@ -0,0 +1,113 @@
+// then waits for a successful connectivity check to the ip.
+func (t *IngressUpgradeTest) Setup(f *framework.Framework) {
+	// jig handles all Kubernetes testing logic
+	jig := framework.NewIngressTestJig(f.ClientSet)
+
+	ns, err := f.CreateNamespace("ingress-upgrade", nil)
+	framework.ExpectNoError(err)
+
+	// gceController handles all cloud testing logic
+	gceController := &framework.GCEIngressController{
+		Ns:     ns.Name,
+		Client: jig.Client,
+		Cloud:  framework.TestContext.CloudConfig,
+	}
+
+	t.gceController = gceController
+	t.jig = jig
+	t.httpClient = framework.BuildInsecureClient(framework.IngressReqTimeout)
+
+	// Allocate a static-ip for the Ingress, this IP is cleaned up via CleanupGCE
+	t.ip = t.gceController.CreateStaticIP(ns.Name)
+
+	// Create a working basic Ingress
+	By(fmt.Sprintf("allocated static ip %v: %v through the GCE cloud provider", ns.Name, t.ip))
+	jig.CreateIngress(filepath.Join(framework.IngressManifestPath, "static-ip"), ns.Name, map[string]string{
+		"kubernetes.io/ingress.global-static-ip-name": ns.Name,
+		"kubernetes.io/ingress.allow-http":            "false",
+	})
+
+	By("waiting for Ingress to come up with ip: " + t.ip)
+	framework.ExpectNoError(framework.PollURL(fmt.Sprintf("https://%v/", t.ip), "", framework.LoadBalancerPollTimeout, jig.PollInterval, t.httpClient, false))
+}
+
+// Test waits for the upgrade to complete, and then verifies
+// with a connectvity check to the loadbalancer ip.
+func (t *IngressUpgradeTest) Test(f *framework.Framework, done <-chan struct{}, upgrade UpgradeType) {
+	switch upgrade {
+	case MasterUpgrade:
+		// Restarting the ingress controller shouldn't disrupt a steady state
+		// Ingress. Restarting the ingress controller and deleting ingresses
+		// while it's down will leak cloud resources, because the ingress
+		// controller doesn't checkpoint to disk.
+		t.verify(f, done, true)
+	default:
+		// Currently ingress gets disrupted across node upgrade, because endpoints
+		// get killed and we don't have any guarantees that 2 nodes don't overlap
+		// their upgrades (even on cloud platforms like GCE, because VM level
+		// rolling upgrades are not Kubernetes aware).
+		t.verify(f, done, false)
+	}
+}
+
+// Teardown cleans up any remaining resources.
+func (t *IngressUpgradeTest) Teardown(f *framework.Framework) {
+	if CurrentGinkgoTestDescription().Failed {
+		framework.DescribeIng(t.gceController.Ns)
+	}
+	if t.jig.Ingress == nil {
+		By("No ingress created, no cleanup necessary")
+		return
+	}
+	By("Deleting ingress")
+	t.jig.DeleteIngress()
+
+	By("Cleaning up cloud resources")
+	framework.CleanupGCE(t.gceController)
+}
+
+func (t *IngressUpgradeTest) verify(f *framework.Framework, done <-chan struct{}, testDuringDisruption bool) {
+	if testDuringDisruption {
+		By("continuously hitting the Ingress IP")
+		framework.ExpectNoError(framework.PollURL(fmt.Sprintf("https://%v/", t.ip), "", framework.LoadBalancerPollTimeout, t.jig.PollInterval, t.httpClient, false))
```

Comment:
After reading your comment below, the problem is that this stops polling after one successful response. What you should do is poll until the the disruption is complete. The service test does this by doing this one-shot polls inside a `wait.Until` call that stops when `done` is closed.